### PR TITLE
Eliminate shared this from std/socket.d

### PR DESCRIPTION
**Pull request by andralex**
This ends up simplifying Posix a lot. It could be simplified even further.

### CODE REVIEW

**Comment by andralex** [comment]
cc @CyberShadow @WalterBright @braddr @wilzbach @jmdavis @Imagix

**Comment by CyberShadow** [comment]
I think we can now get rid of `getnameinfo` being conditional by now. It was starting to lose sense when it was introduced in 2011, and today we probably don't support anything that old because of other reasons.

**Comment by CyberShadow** [comment]
BTW, I'm skeptical about replacing `static ~this` with `atexit`. 

- Is `atexit` thread-safe? `man` says it's thread-safe but what about TLS variables? Will the function specified to `atexit` be called in the same thread as the code calling `atexit`?
- Do the semantics match up correctly? Such as thread/process distinctions, and unloading of DLLs?
- `atexit` seems to have a limit of 32 functions (`ATEXIT_MAX`).
- Not sure about the merits of replacing a D feature with a C feature.

**Comment by andralex** [comment]
>I think we can now get rid of getnameinfo being conditional by now. 

OK, I'll look into that.

I think we're good with atexit only as long as we only mess with other C libraries.

**Comment by andralex** [comment]
@CyberShadow just to make sure: on Posix getnameinfo, getaddrinfo, and freeaddrinfo can never be null weak symbols. Correct?

**Comment by CyberShadow** [comment]
Needs updates to documentation (e.g. lots of references to `SocketFeatureException` don't apply any more).

**Comment by andralex** [comment]
@CyberShadow done.

All - I notice that `setKeepAlive` throws depending on a statically known condition, which seems kinda silly. Should we just `static assert(false)` in that case? Then we can remove `SocketFeatureException` altogether.

**Comment by CyberShadow** [comment]
Nice work. Would be good to see the project tester's results for this one (so no auto-merge so far).

> All - I notice that setKeepAlive throws depending on a statically known condition, which seems kinda silly.

Yeah, that's a bit of a Java-ism.

> Should we just static assert(false) in that case?

It would mean making the function a template. Currently it's virtual due to virtual-by-default, and because of that Socket methods are all pretty sensitive to breakage (see [issue 16514](https://issues.dlang.org/show_bug.cgi?id=16514)).

I guess there's also the small risk that there is code out there that already has `setKeepAlive` behind a regular `if` (not static).

**Comment by andralex** [comment]
Cool, so I guess this PR is good to go. @CyberShadow ?

File: std/socket.d [hunk 1]
```diff
@@ -328,33 +342,11 @@ shared static this() @system
...
-    else version(Posix)
-    {
-        getnameinfoPointer = &getnameinfo;
-        getaddrinfoPointer = &getaddrinfo;
-        freeaddrinfoPointer = &freeaddrinfo;
-    }
-}
-
+        // Now that we called WSAStartup, make sure we clean up, too.
+        import core.stdc.stdlib;
```
Comment (CyberShadow) [inline std/socket.d]: `import core.stdc.stdlib : atexit;` as per CircleCI

**Comment by CyberShadow** [comment]
Still lots of errors in the autotester.

File: std/socket.d (hunk 1 above)
Comment (andralex) [inline std/socket.d]: Urgh. Not a fan of this rule. I recall we introduced it to compensate for hijacking, which we since fixed.

**Comment by andralex** [comment]
OK, pushed fixes to the autotester problems. Flying blindly here...

**Comment by andralex** [comment]
moar fixes

**Comment by andralex** [comment]
Hmmm, so now we have a different problem:

    Error 42: Symbol Undefined _getnameinfo@28

It seems this is caused by switching from dynamic loading to statically linking. What library do we need to link with, and what's the cmdline syntax? Thx!

**Comment by CyberShadow** [comment]
You'll need to update the import libraries... :/

There are some definition files in druntime/def, but none for winsock. The best way may be to have the DMC import libraries updated. @WalterBright?

**Comment by WalterBright** [comment]
> What library do we need to link with, and what's the cmdline syntax?

Use `grep`, the all-purpose tool:
```
grep getnameinfo \dm\lib\*.lib
```
Unfortunately, no hits. Trying:
```
grep getnameinfo \dm\bin\*.dll
```
No hits there, either. I don't know where it is supposed to be; it isn't in the DMC distribution.

**Comment by CyberShadow** [comment]
> I don't know where it is supposed to be; it isn't in the DMC distribution.

You need to add getnameinfo and getaddrinfo (possibly others?) to ws2_32.lib.

**Comment by WalterBright** [comment]
> You need to add getnameinfo and getaddrinfo (possibly others?) to ws2_32.lib.

The dm\bin\coffimplib.exe program can be used to convert 32 bit MS-Coff files to OMF.

**Comment by andralex** [comment]
Reverted the pull that causes problems on Windows.

**Comment by andralex** [comment]
This should be good to go subject to the autotester passing.

**Comment by andralex** [comment]
I'm throwing away this entire pile of dung and start anew.

**Comment by andralex** [comment]
Redone in https://github.com/dlang/phobos/pull/5813

**Pull request closed without merging**